### DISCUSSION TITLE
User IDs as strings instead of UUIDs for external auth?

### DISCUSSION THREAD

**Discussion by colinclerk** [supabase/supabase#4954] [Questions] [answered]
Hello - I work for clerk.dev (an authentication service) and we're working to improve our compatibility with Supabase.

A key challenge is that our User IDs are strings (prefixed KSUIDs) instead of UUIDs.  This is breaking the `auth.uid()` helper which expects a UUID in the `sub` claim.  In particular, we think line 100 is throwing errors:
https://github.com/supabase/supabase/blob/c8d910ac7a21ea468822f9e28a691f1687f60a58/docker/volumes/db/init/01-auth-schema.sql#L92-L101

Is there a recommended way to get past this?  We'd like to better understand the different options.

Currently, our guide omits the `sub` claim and uses a `userId` claim instead.  We then ask developers to define a separate function to grab the user ID.  This is unfortunate since it means developers cannot use `auth.uid()` and it's not a great developer experience overall.

We've had three other ideas:
* We can submit a PR to cast auth.uid() to text if casting to a uuid fails.  This is our preference if it would be accepted.
* We can try to instruct developers to override auth.uid().  We haven't explored this as an option because we are concerned with forward compatibility through supabase upgrades.  Is this a valid concern?
* We can start generating UUIDs for each user, and put that into the sub claim instead.  We're not huge fans of this since we'd be giving 2 IDs to each user.

**Comment by arcavid** [comment]
It'd be just amazing to have the simple `auth.uid()` work.
I wonder if this feature is being worked on since this issue is more than two months old.

**Comment by FranckPoingt** [comment]
Any update on the matter? The provided workaround doesn't work...

**Comment by JonathanPicques** [comment]
Maybe creating a new policy (like `auth.ouruid()`) could help?
See https://github.com/supabase/supabase/discussions/1849

**Comment by alex-galey** [comment]
The [suggested workaround](https://clerk.dev/docs/integration/supabase) didn't work for me and this is a deal breaker not to be able to implement RLS

- I defined a custom function in postgres `requesting_user_id()`
`select nullif(current_setting('request.jwt.claims', true)::json->>'sub', '')::text;`

- The function is seen on project/xxx/database/functions

- I defined an RLS policy (RLS is active on the table ofc) for the users only to be able to get their own data :
```
CREATE POLICY "Pol 1" 
	ON public.user USING (
		auth.role() = 'authenticated'::text
	) WITH CHECK (
		requesting_user_id() = auth_id
	);
```

- I used the [Clerk + Supabase template](https://github.com/clerkinc/clerk-supabase-starter) as a strating point

- I was able to insert a row with user_id : 
```
const token = await getToken({ template: 'supabase-user-dashboard' });
supabase.auth.setAuth(token);
await supabase.from('user').upsert({
   auth_id: userId,
   firstname: 'Alex',
   lastname: 'G'
   }, { returning: 'minimal' })
```

userId is coming from ```const { userId, getToken } = useAuth();```

- I manually added another row with non-matching auth_id and queried the table again, which **now returns the other row too** : 
```
const token = await getToken({ template: 'supabase-user-dashboard' });
supabase.auth.setAuth(token);
const { data } = await supabase.from('user').select();
```

Any update on when the native auth.uid() api call would be working ?

**Comment by steve-chavez** [comment]
Hey @colinclerk, sorry for the late reply here.

> We can submit a PR to cast auth.uid() to text if casting to a uuid fails. This is our preference if it would be accepted.

Can you show a snippet of how would you do that? Not sure if possible since the function signature would have to change as well(`returns text`?)


---

As suggested by JonathanPicques, another option would be to add a custom funcion to the `auth` schema. Maybe an `auth.external_id()`(it could take a key parameter in case the id doesn't come from the `sub` claim)

**Comment by mypalvikram** [comment]
Hello, just wanted to check if there have been any updates on this or if using the custom function is still the way to go. Thanks!

**Comment by rdylina** [comment]
@colinclerk I'm also curious if there has been any progress on this topic.

**Comment by sohcfst** [comment]
Would love integrate clerk and supabase in my project but coming across several snags as I'm not great with understanding RLS or SQL in general ):

**Comment by tonyabracadabra** [comment]
Also looking for this!

**Comment by kenmwangi** [comment]
Is there any progress on this issue? Seem to take a long time to resolve

**Comment by castroalves** [comment]
Any news on that?

**Comment by mansueli** [comment, marked as answer]
First, you'll need to drop the FK from the storage.objects table:

```sql
ALTER TABLE storage.objects 
DROP CONSTRAINT objects_owner_fkey;
```

Here's a possible workaround for converting any text into uuid by hashing it, then formatting as UUID:

```sql
CREATE OR REPLACE FUNCTION convert_to_uuid(input_value text)
RETURNS uuid AS $$
DECLARE
  hash_hex text;
BEGIN
  -- Return null if input_value is null or an empty string
  IF input_value IS NULL OR NULLIF(input_value, '') IS NULL THEN
    RETURN NULL;
  END IF;
  hash_hex := substring(encode(digest(input_value::bytea, 'sha512'), 'hex'), 1, 36);
  RETURN (left(hash_hex, 8) || '-' || right(hash_hex, 4) || '-4' || right(hash_hex, 3) || '-a' || right(hash_hex, 3) || '-' || right(hash_hex, 12))::uuid;
END;
$$ LANGUAGE plpgsql;
````

Then, you could edit the [clerk](https://supabase.com/docs/guides/integrations/clerk#access-user-id-in-rls-policies)'s requesting_user_id function to use this instead:

```sql
create or replace function requesting_user_id()
returns uuid
language sql stable
as $$
  select convert_to_uuid(current_setting('request.jwt.claims', true)::json->>'sub');
$$;
```

**Reply by novemberkilo** [reply]
Thanks for this @mansueli 

I was able to follow your suggestions here but with some modification. For the purposes of storing the generated uuid in a `user_id` column, I tried to use your `convert_to_uuid` function with something like

```
ADD COLUMN user_id uuid GENERATED ALWAYS AS (convert_to_uuid(clerk_user_id)) STORED;
```

but couldn't because of the error `generation expression is not immutable` -- if you do `\df+ convert_to_uuid` in psql you will see that this function is `volatile`

I was however, able to use postgres' [uuid functions](https://www.postgresql.org/docs/current/uuid-ossp.html) to accomplish this and get RLS working via a corresponding change to `requesting_user_id`

fwiw I used `generate_uuid_v5` in place of `convert_to_uuid` successfully.

**Comment by tonyboylehub** [comment]
Can this be picked up and looked at again for proper intergration? You guys are "partners" according to you (Supabase) per your site and claim to be intergrated with each other but this still isn't resolved after 10 months nearly.

Storage is still broken and It's not a full intergration of service currently, or it at least needs to be noted in docs from both sides that there is "limited" intergration at least. That's 10 months of devs picking these two as a potential service to find out they are not fully integrated quite a way into a build/test.

Edit: I thought this originally brought up in Jan 2023 (only read the Jan part intially), but it was Jan 2022 this was brought to attention.

1 year 10 months this has been a thing.

**Comment by skorfmann** [comment]
```
ALTER TABLE storage.objects 
DROP CONSTRAINT objects_owner_fkey;
```

shouldn't be required anymore, see https://github.com/supabase/storage-api/issues/276#issuecomment-1596745720

**Reply by tonyboylehub** [reply]
When ever your try and use Clerks JWT Auth for storage with signed URLs it still checks for a UUID in `sub` of the JWT and fails because it can't convert it to type of UUID. I don't believe it even makes it as far as the table.

**Reply by skorfmann** [reply]
So everything which is relying on UUID won't work, correct? e.g. [user based permissions](https://supabase.com/docs/guides/storage/access-control#allowing-individual-access-to-a-file) for storage since `owner` is `uuid`

```
create policy "Individual user Access"
on storage.objects for select
to authenticated
using ( auth.uid() = owner );
```

anything else which is affected by this?

**Reply by tonyboylehub** [reply]
I partly missed some of the issue I think, it's with the [ createSignedUploadUrl](https://supabase.com/docs/reference/javascript/storage-from-createsigneduploadurl?example=create-signed-upload-url).

By default the storage.objects `owner` was a UUID, I've sinced used the sql editor to drop that column and create a new one as a texr but no luck still and still produces the same error.

In the Clerk JWT the user id is stored in `sub` of the token.

Then checking the sub of the token for the userID, which I've currently been using for all database relations. 

 `error: StorageApiError: invalid input syntax for type uuid: "user_xxxxxxxxxxxxxxxxxxxxxxxxxxxx"`
 I can also see this log in the SQL logs as well.
 
 I thought I had removed all the instances of UUID but clearly I haven't and something else is at play too.
 
 I'm also a little aprehensive of changing to much in the objects table, as mentioned above and in multiple threads that the tables could be overwritten by Supabase it seems during an update/migration.
 
 I've tried all the suggestions so far from policies, addtional sql functions, table editing/column renaming.

**Comment by Eliot6001** [comment]
Maybe turn it into a separate small dependency that you add only if you're using the combo of "Supabase + clerk" which would include a bunch of functions such as `clerkbase.auth()` ?

**Comment by clxyder** [comment]
Any updates?

**Comment by ivan-robert** [comment]
something is "marked as an answer" but does not answer at all /-_-/  In the storage query the id is directly extracted from the "sub", the fucntion"requesting_user_id" is not even called

**Comment by brysonjones** [comment]
Any thoughts on how folks should proceed here, @colinclerk @mansueli?

**Comment by bruno-at-breezy** [comment]
Any updates?

**Comment by aishik-pyne** [comment]
I have been facing this exact same problem. 
Currently, Supabase with Clerk is working fine for most of the tables which are generated by me, as I defined the RLS policy to use the `owner_id::text` or `user_id::text` as the Clerk ID from `auth.jwt() ->> 'sub'`.

However, when I try to do the same for storage bucket policies, it fails. The Storage.objects table has two columns: owner and owner ID. Owner is a UUID and owner ID is text. Although the text field is being populated on upload, while retrieving the policy fails to execute and gives a mismatch error, saying that the user ID, which should have been a UUID, is coming as text. 

<img width="1381" height="225" alt="image" src="https://github.com/user-attachments/assets/b1386a41-dd5b-44f8-b0ad-20da0af2b85c" />


How do I make the RLS policy work with the owner ID column only?

**Comment by solvionsolutions** [comment]
If anyone lands here in 2026, there's a much cleaner path now than the `convert_to_uuid()` hashing trick — and worth flagging that the SHA-512 approach still quietly breaks Storage, since Storage policies read the raw `sub` claim directly and never call your custom function. Now that Clerk ships a native third-party auth integration for Supabase, you don't need to force those KSUIDs into UUIDs at all. Register Clerk as a third-party auth provider in your Supabase project (it passes the JWT through so `auth.jwt()->>'sub'` returns Clerk's real string id), keep your `user_id` columns as `text` instead of `uuid`, and just compare them directly:

```sql
create policy "Users manage their own rows"
on public.notes
for all
using  (user_id = auth.jwt()->>'sub')
with check (user_id = auth.jwt()->>'sub');
```

No hashing, no lossy casts, and it behaves identically for Storage buckets because those policies read the same claim. The originally accepted answer was the right call back in 2023 — this is really just the modern way to do it now that the native integration exists.

I wrote up the full breakdown here: https://www.guardlayer.io/blog/clerk-supabase-rls
